src/System.Management.Automation/engine/parser/DynamicKeyword.cs

```diff
@@ -0,0 +1,2759 @@
+        public string ResourceName { get; set; } = DefaultResourceName;
+
+        /// <summary>
+        /// Indicates whether a keyword uses a marshalled call or is a direct function call, for DSC node keywords
+        /// </summary>
+        public bool DirectCall { get; set; } = DefaultIsDirectCall;
+
+        /// <summary>
+        /// Indicates that the keyword should not be added to the AST (and therefore should do nothing at runtime) if true
+        /// </summary>
+        public bool MetaStatement { get; set; } = DefaultIsMetaStatement;
+    }
+
+    /// <summary>
+    /// Attribute defining a class property as a PowerShell dynamic keyword parameter
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Property)]
+    public class KeywordParameterAttribute : ParsingBaseAttribute
+    {
+        /// <summary>
+        /// Parameters are not mandatory by default
+        /// </summary>
+        public const bool NotMandatory = false;
+
+        /// <summary>
+        /// Construct a KeywordParameterAttribute with default values
+        /// </summary>
+        public KeywordParameterAttribute()
+        {
+        }
+
+        /// <summary>
+        /// Specifies whether this parameter is mandatory
+        /// </summary>
+        public bool Mandatory { get; set; } = NotMandatory;
+    }
+
+    /// <summary>
+    /// Attribute defining a class property as a PowerShell dynamic keyword property (for hashtable keywords)
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Property)]
+    public class KeywordPropertyAttribute : ParsingBaseAttribute
+    {
+        /// <summary>
+        /// Properties are not mandatory by default
+        /// </summary>
+        public const bool NotMandatory = false;
+
+        /// <summary>
+        /// Construct a KeywordPropertyAttribute with default values
+        /// </summary>
+        public KeywordPropertyAttribute()
+        {
+        }
+
+        /// <summary>
+        /// Specifies whether this property is mandatory
+        /// </summary>
+        public bool Mandatory { get; set; } = NotMandatory;
+    }
+
+    #endregion
+}
+
+// The remainder of the DynamicKeyword functionality is parser specific,
+// and is placed in the Automation.Language namespace because DynamicKeyword is there
+
+namespace System.Management.Automation.Language
+{
+    #region Dynamic Keyword Parser Datastructures
+
+    #region Dynamic Keyword Namespacing
+
+    /// <summary>
+    /// Defines a scoped namespace for Dynamic Keywords. This class is intended to both
+    /// encapsulate nested dynamic keyword scoping, and be the first step toward making
+    /// DynamicKeyword storage parser local instead of thread static
+    /// </summary>
+    internal class DynamicKeywordNamespace
+    {
+        /// <summary>
+        /// Keeps track of the DynamicKeyword scope by storing enclosing
+        /// DynamicKeywords on the stack
+        /// </summary>
+        private Stack<DynamicKeyword> DynamicKeywordScope
```

Comment:
So namespaces don't nest - a scope is a `DynamicKeyword`?

It might be useful to introduce a scope instead of hanging the scope-like state off a `DynamicKeyword` - this way maybe there is a way to unify the notion of `Global` and nested keywords.